# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1362140298 18000
#      Fri Mar 01 07:18:18 2013 -0500
# Node ID 22bb3aa9f025c1e148ecb3d78bd29b32ec48cae7
# Parent  cb80b1d062b1f448884f5aedd0127951c98b2f0c
make push parser interface partially work

* oct-parse.yy: Use USE_OCTAVE_PUSH_PARSER, not USE_PUSH_PARSER.
(octave_parser::~octave_parser): Cast parser state to yypstate*.
(octave_parser::run): Simplify.  If USE_OCTAVE_PUSH_PARSER, use push
parser interface, but still require that the lexer is in charge of
gathering input.

diff --git a/libinterp/parse-tree/oct-parse.yy b/libinterp/parse-tree/oct-parse.yy
--- a/libinterp/parse-tree/oct-parse.yy
+++ b/libinterp/parse-tree/oct-parse.yy
@@ -1477,66 +1477,49 @@ opt_comma       : // empty
 static void
 yyerror (octave_parser *curr_parser, const char *s)
 {
   curr_parser->bison_error (s);
 }
 
 octave_parser::~octave_parser (void)
 {
-#if defined (USE_PUSH_PARSER)
-  yypstate_delete (parser_state);
+#if defined (OCTAVE_USE_PUSH_PARSER)
+  yypstate_delete (static_cast<yypstate *> (parser_state));
 #endif
 
 delete curr_lexer;
 }
 void octave_parser::init (void)
 {
-#if defined (USE_PUSH_PARSER)
+#if defined (OCTAVE_USE_PUSH_PARSER)
   parser_state = yypstate_new ();
 #endif
 
   CURR_LEXER = curr_lexer;
 }
 
 int
 octave_parser::run (void)
 {
   int status = 0;
 
-#if defined (USE_PUSH_PARSER)
-
-  for (;;)
+#if defined (OCTAVE_USE_PUSH_PARSER)
+
+  do
     {
-      unwind_protect frame;
-
-      frame.protect_var (current_input_line);
-
-      bool eof = false;
-
-      get_user_input (eof);
-
-      do
-        {
-          octave_char = eof ? END_OF_INPUT : octave_lex ();
-
-          if (octave_char == 0)
-            {
-              // Attempt to get more input.
-              status = -1;
-              break;
-            }
-
-          status = octave_push_parse (pstate);
-        }
-      while (status == YYPUSH_MORE);
-
-      if (status >= 0)
-        break;
+      YYSTYPE lval;
+
+      int token = octave_lex (&lval, scanner);
+
+      yypstate *pstate = static_cast<yypstate *> (parser_state);
+
+      status = octave_push_parse (pstate, token, &lval, this);
     }
+  while (status == YYPUSH_MORE);
 
 #else
 
   status = octave_parse (this);
 
 #endif
 
   return status;
